extensions/notebook/src/jupyter/jupyterNotebookProvider.ts

```diff
@@ -46,15 +51,52 @@ export class JupyterNotebookProvider implements nb.NotebookProvider {
 			// As this is a notification method, will skip throwing an error here
 			return;
 		}
-		let uriString = notebookUri.toString();
-		let manager = this.managerTracker.get(uriString);
+		let baseFolder = this.transformToBaseFolder(notebookUri.fsPath.toString());
+		let manager = this.managerTracker.get(baseFolder);
 		if (manager) {
-			this.managerTracker.delete(uriString);
-			manager.dispose();
+			let sessionManager = (manager.sessionManager as JupyterSessionManager);
+			let session = sessionManager.listRunning().find(e => e.path === notebookUri.fsPath);
+			if (session) {
+				manager.sessionManager.shutdown(session.id);
+			}
+			if (sessionManager.listRunning().length === 0) {
+				const FiveMinutesInMs = 5 * 60 * 1000;
+				setTimeout(() => {
+					if (sessionManager.listRunning().length === 0) {
+						this.managerTracker.delete(baseFolder);
+						manager.dispose();
+					}
+				}, FiveMinutesInMs);
+			}
 		}
 	}
 
 	public get standardKernels(): nb.IStandardKernel[] {
 		return [];
 	}
+
+	private transformToBaseFolder(notebookPath: string): string {
+		let parsedPath = path.parse(notebookPath);
+		let userHome = utils.getUserHome();
+		let relativePathstrUserHome = path.relative(notebookPath, userHome);
+		if (notebookPath === '.' || relativePathstrUserHome.endsWith('..') || relativePathstrUserHome === '') {
+			// If you don't match the notebookPath's casing for drive letters,
```

Comment:
You're never returning anything if this statement is false. (and if you end up having undefined be a possible return value make sure to update the return type to string | undefined)